code/modules/jobs/job_exp.dm

```diff
@@ -0,0 +1,274 @@
+			message_admins("SQL ERROR during exp_update_client update. Error : \[[err]\]\n")
+			return
+
+
+
+/client/proc/update_exp_list(var/minutes, var/announce_changes = 0)
+	if(!src ||!ckey || !config.use_exp_tracking)
+		return
+
+	var/list/play_records = prefs.exp
+	var/list/old_records = prefs.exp.Copy()
+
+	for(var/rtype in GLOB.exp_jobsmap)
+		if(!play_records[rtype])
+			play_records[rtype] = 0
+	if(mob.stat == CONSCIOUS && mob.mind.assigned_role)
+		play_records[EXP_TYPE_LIVING] += minutes
+		if(announce_changes)
+			to_chat(mob,"<span class='notice'>You got: [minutes] Living EXP!")
+		for(var/category in GLOB.exp_jobsmap)
+			if(GLOB.exp_jobsmap[category]["titles"])
+				if(mob.mind.assigned_role in GLOB.exp_jobsmap[category]["titles"])
+					play_records[category] += minutes
+					if(announce_changes)
+						to_chat(mob,"<span class='notice'>You got: [minutes] [category] EXP!")
+		if(mob.mind.special_role)
+			play_records[EXP_TYPE_SPECIAL] += minutes
+			if(announce_changes)
+				to_chat(mob,"<span class='notice'>You got: [minutes] Special EXP!")
+	else if(isobserver(mob))
+		play_records[EXP_TYPE_GHOST] += minutes
+		if(announce_changes)
+			to_chat(mob,"<span class='notice'>You got: [minutes] Ghost EXP!")
+	else if(minutes)	//Let "refresh" checks go through
+		return
+
+	for(var/jtype in play_records)
+		if(play_records[jtype] != old_records[jtype])
+			LAZYINITLIST(GLOB.exp_to_update[ckey])
+			LAZYINITLIST(GLOB.exp_to_update[ckey][jtype])
+			GLOB.exp_to_update[ckey][jtype] = play_records[jtype]
+	prefs.exp = play_records
+
+
+
+
+//resets a client's exp to what was in the db.
+/client/proc/set_exp_from_db()
+	if(!src ||!ckey || !config.use_exp_tracking)
+		return
+	if(!SSdbcore.Connect())
+		return -1
+	var/datum/DBQuery/exp_read = SSdbcore.NewQuery("SELECT job, minutes FROM [format_table_name("role_time")] WHERE ckey = '[sanitizeSQL(ckey)]'")
+	if(!exp_read.Execute())
+		var/err = exp_read.ErrorMsg()
+		log_game("SQL ERROR during exp_update_client read. Error : \[[err]\]\n")
+		message_admins("SQL ERROR during exp_update_client read. Error : \[[err]\]\n")
+		return
+	var/list/play_records = list()
+	while(exp_read.NextRow())
+		play_records[exp_read.item[1]] = text2num(exp_read.item[2])
+
+	for(var/rtype in GLOB.exp_jobsmap)
+		if(!play_records[rtype])
+			play_records[rtype] = 0
+
+	prefs.exp = play_records
+
+//writes everything in the exp_to_update list to the db
+/proc/update_exp_db()
+	if(!config.use_exp_tracking)
+		return
+	if(!SSdbcore.Connect())
+		return -1
+	for(var/keys in GLOB.exp_to_update)
+		var/keyname = keys
+		for(var/jtype in GLOB.exp_to_update[keyname])
+			var jobname = jtype
+			var time = GLOB.exp_to_update[keyname][jtype]
+			var/datum/DBQuery/update_query = SSdbcore.NewQuery("INSERT INTO [format_table_name("role_time")] (`ckey`, `job`, `minutes`) VALUES ('[sanitizeSQL(keyname)]', '[sanitizeSQL(jobname)]', '[sanitizeSQL(time)]') ON DUPLICATE KEY UPDATE minutes = '[sanitizeSQL(time)]'")
```

Comment:
if they play on server 1, die after 10 minutes, then play on server 2 and get an hour of play time, all of that will be lost. if server 1 updates its round after server 2.
You can not assume your stored data is still good.

The best way would be to do this when you update everybody's exp, as += queries

Also, you can merge this into one big query easy, rather then do it in a loop. i'll make a generic function for doing so and get back to you.

when you do ON DUPLICATE KEY UPDATE, you should also always use `column = VALUES(column)` so that it pulls from the values and you don't have to send the same data twice, it also helps when you do multiple row inserts since it just pulls the data that would have been inserted.